src/redis-cli.c

```diff
@@ -794,15 +795,24 @@ static int cliAuth(redisContext *ctx, char *user, char *auth) {
     return REDIS_ERR;
 }
 
-/* Send SELECT dbnum to the server */
+/* Send SELECT input_dbnum to the server */
 static int cliSelect(void) {
     redisReply *reply;
-    if (config.dbnum == 0) return REDIS_OK;
+    if (config.input_dbnum == config.dbnum) return REDIS_OK;
 
-    reply = redisCommand(context,"SELECT %d",config.dbnum);
+    reply = redisCommand(context,"SELECT %d",config.input_dbnum);
     if (reply != NULL) {
         int result = REDIS_OK;
-        if (reply->type == REDIS_REPLY_ERROR) result = REDIS_ERR;
+        if (reply->type == REDIS_REPLY_ERROR) {
+            result = REDIS_ERR;
+            fprintf(stderr,"SELECT %d failed: %s\n",config.input_dbnum,reply->str);
+            if (strncmp(reply->str, "NOAUTH", 6)) {
+                config.input_dbnum = config.dbnum;
```

Comment:
why do we reset `config.input_dbnum` here?
not sure we ever wanna reset it.
i think the only places that should write to it are the `-n` option, and maybe when SELECT command was executed successfully, i.e. this code:
```c
            /* Store database number when SELECT was successfully executed. */
            if (!strcasecmp(command,"select") && argc == 2 && 
                config.last_cmd_type != REDIS_REPLY_ERROR) 
```
but why do the other places write to that var, and specifically this one (when AUTH fails)?

Reply:
If users select failed because of `NOAUTH`, next time when `AUTH` is executed, users can select right dbnum.
If failed by other reasons, every time users reconnect and `SELECT`, I think this `SELECT` will always fails.

Why  write to that var? If something wrong happens with the connection, the connection will automatically reconnect.
If `config.input_dbnum` is different with `config.dbnum`, the new selected dbnum will be `config.input_dbnum`. I think we need `config.dbnum`.

Reply:
ok, i see. but i'm not sure we wanna distinguish between `NOAUTH` and other errors.
what you said is right about `SELECT is not allowed in cluster mode` and `DB index is out of range`, but maybe not on connection errors or NOPERM etc.
i think i'd rather let it fail every time.

Reply:
i've just tested this branch,
if i use an invalid db, the prompt is wrong:
```
$ src/redis-server --requirepass 123 &

$ redis-cli  -n 20 -a 123
Warning: Using a password with '-a' or '-u' option on the command line interface may not be safe.
127.0.0.1:6379[20]> 
127.0.0.1:6379[20]> client info
"id=4 addr=127.0.0.1:44628 laddr=127.0.0.1:6379 fd=8 name= age=3 idle=0 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=26 qbuf-free=32744 argv-mem=10 obl=0 oll=0 omem=0 tot-mem=49794 events=r cmd=client user=default redir=-1\n"
```
please look into it.

Reply:
Everytime try to reconnect and fail is OK, because reconnect rarely happens.

Reply:
@oranagra It's like you are using an old redis-cli.
Should you use `src/redis-cli  -n 20 -a 123`?

Reply:
ohh, i'm an idiot!

Reply:
LOL! using the wrong redis-cli. when the redis-cli is exactly what i came to test..
smells as if i just made the transition from Windows (which i haven't used in at least 10 years!)

Reply:
I found other two bugs. One is easy to fix. I fix it in the new commit.
Another is a little hard to fix and is not easy to happen.
```
127.0.0.1:6383> multi
OK
127.0.0.1:6383(TX)> select 20
QUEUED
127.0.0.1:6383[20](TX)> exec
1) (error) ERR DB index is out of range
127.0.0.1:6383[20]>
```

Reply:
i think i'm willing to let this one slide. it's a very rare and unlikely edge case, and in anyway this feature of showing the selected db is just an attempt to assist the user, not a critical feature.
the only easy way to fix it AFAIK is to always call CLIENT INFO after exec succeeds, but i don't like the fact redis-cli implicitly sends many commands to to redis.
we do have a COMMAND command at startup, and we may re-try SELECT after the user sends a manual AUTH, but i don't like to send a CLIENT INFO after each EXEC.

if you agree, and can't think of any other easy way to solve it, i think we can merge this as it is.

Reply:
Agree. So I left it. The fix will not be easy.